lib/Controller/RoomController.php

```diff
@@ -1316,7 +1301,23 @@ public function addParticipantToRoom(string $newParticipant, string $source = 'u
 			return new DataResponse([], Http::STATUS_BAD_REQUEST);
 		}
 
-		return new DataResponse();
+		// attempt adding the listed users to the room
+		// existing users with USER_SELF_JOINED will get converted to regular USER participants
+		foreach ($participantsToAdd as $participantToAdd) {
+			$existingParticipant = $participantsByUserId[$participantToAdd['actorId']] ?? null;
+
+			if ($existingParticipant !== null) {
+				if ($existingParticipant->getAttendee()->getParticipantType() !== Participant::USER_SELF_JOINED) {
+					// user is already a regular participant, skip
+					continue;
+				}
+				$this->participantService->updateParticipantType($this->room, $existingParticipant, Participant::USER);
+			} else {
+				$this->participantService->addUsers($this->room, [$participantToAdd]);
```

Comment:
can we still group them? otherwise this will trigger too many events when adding a group with a lot of users.

Reply:
ah! indeed, I was assuming the function was simply looping over each entry on its own.
I missed the part about events, thanks for pointing out.

I'll adjust the impl to do the "addUsers" in a single batch.

Hmmm, then maybe we should do the same for updating the participant types as it has its own event.

Reply:
Feels a bit overkill to create a new ModifyParticipantsEvent (plural) for a corner case.

I guess I'll only batch the `addUsers` one and leave the other be.

Reply:
> I'll only batch the addUsers one and leave the other be.

yes